### PR TITLE
Retrofit the release workflow to be useful for doing a pre-release

### PR DESCRIPTION
**Pull request by ariya**
Once this PR is merged, there will be a manually-triggered workflow which does the following, for each OSS and EE edition, to facilitate testing both the Uberjars and container images (before a release):

1. Build Uberjars, stash the Uberjars as downloadable artifacts
2. Check the health check when launching each Uberjar with Java 11 and 17
3. Put the Uberjar in the container image, push it to DockerHub
4. Verify the image by doing `docker pull`

Step 3 and 4 are carried out in isolation, thereby allowing the full round-trip (`docker push` and `docker pull`) verification.

Testing this is easier done in a fork (e.g. in my case, [github.com/ariya/metabase](https://github.com/ariya/metabase)). The workflow can be triggered by supplying the commit hash which we want to check:

image

The visualization of all the jobs being executed thereafter:

image

The container image is also available in Docker Hub (shown here only for `metabase-enterprise`, but `metabase` the OSS edition is also available):

image

### CODE REVIEW

File: .github/workflows/pre-release.yml
Comment (ariya) [inline .github/workflows/pre-release.yml:25]: Manual commit switch seems to be more reliable, based on my testing.

File: .github/workflows/pre-release.yml
Comment (ariya) [inline .github/workflows/pre-release.yml:86]: Bump the action version, to avoid the nagging of outdated Node.js v12 action.

File: .github/workflows/pre-release.yml
Comment (ariya) [inline .github/workflows/pre-release.yml:90]: idem

File: .github/workflows/pre-release.yml
Comment (ariya) [inline .github/workflows/pre-release.yml:115]: `github.repository_owner` is used here so that things are also working correctly in a fork (and therefore permits easy testing of this commit/PR).

File: .github/workflows/pre-release.yml
Comment (nemanjaglumac) [reply .github/workflows/pre-release.yml:25]: More reliable in what sense?
Did you notice failures when using actions/checkout?

**Review by nemanjaglumac — APPROVED** [review]
Works for me :)
Great and thanks for working on this!

**Review by nemanjaglumac — CHANGES_REQUESTED** [review]
Actually, wait.
There are a few improvement that we can still make.

Sorry for going back and forth but I didn't notice something originally.

File: .github/workflows/pre-release.yml
Comment (nemanjaglumac) [inline .github/workflows/pre-release.yml:127]: WDYT if we have `verify-docker-pull` only as a **step** in a previous job, rather than having to spin up new container for a completely new **job**?

That way we can save some resources and we can re-use the conditional logic re: `DOCKERHUB_REPO`. That could dry the code up significantly.

File: .github/workflows/pre-release.yml
Comment (ariya) [reply .github/workflows/pre-release.yml:25]: I tried using `ref` again, it consistently failed on me (not sure what I'm missing here):

image

File: .github/workflows/pre-release.yml
Comment (ariya) [reply .github/workflows/pre-release.yml:127]: If we do that, the `docker pull` will only pull from its local cache (since the container image was constructed a few seconds) and not from Dockerhub.
The reason for a separate job is so that the `push` and `pull` are isolated (they don't share the execution environment).

Comment (nemanjaglumac) [reply .github/workflows/pre-release.yml:127]: And environment variables are not persisted between jobs, right?

Comment (ariya) [reply .github/workflows/pre-release.yml:127]: The env variables can be scoped to the entire workflow only if they are defined statically at the top level.

TBH I personally wouldn't worry much about duplication for now, since at some point it can be abstracted away (the case with our custom actions e.g. `prepare-frontend` etc).

File: .github/workflows/pre-release.yml
Comment (nemanjaglumac) [reply .github/workflows/pre-release.yml:25]: Oh, it **might be** because you're not supplying the full SHA?
In all of my testing, I was always copying the full SHA and used it to trigger the workflow.

This is of secondary importance.
I'm happy to use either.

Comment (nemanjaglumac) [reply .github/workflows/pre-release.yml:25]: IF that's the reason why, maybe it's not even bad to use that mechanism as an enforcing function because the SHA itself is going to be used in an image name.

That way, we can avoid having multiple images for the same commit. E.g.
- `metabase/metabase: 3e518549f803ff050fe6ae43ddb6a003908b18a0`
- `metabase/metabase: 3e5185`
- `metabase/metabase: 3e518549f803ff050fe6ae43ddb6a0039`

Comment (nemanjaglumac) [reply .github/workflows/pre-release.yml:25]: ^ Either that or we need to have an additional logic that validates the SHA and possibly truncates it to a minimum required length, regardless of the input provided?

**Review by nemanjaglumac — APPROVED** [review]
Approving (again) just to unblock the PR.
But, since I ended up contributing, please wait for at least one more reviewer to approve before merging.

File: .github/workflows/pre-release.yml
Comment (ariya) [reply .github/workflows/pre-release.yml:25]: That did the trick, thank you!
It's very strange since the canonical `git` can work with any length of SHA1 (as long as it's unique).

Comment (ariya) [reply .github/workflows/pre-release.yml:25]: I like the idea of limiting the length of the SHA1 hash to just 8 characters (more than sufficient, for a while).

Comment (nemanjaglumac) [reply .github/workflows/pre-release.yml]: ^ https://github.com/metabase/metabase/pull/26597/commits/0e73ec3699f7e32c8e32338d6d6c78106accc3a4 should do the trick.

And I switched back to implicitly enforcing full SHA-1 as an input in https://github.com/metabase/metabase/pull/26597/commits/3039a30d544564ce0720f8a709549de9858c67d4.

Feel free to change/adapt any part of it as you see fit.

**Review by flamber — APPROVED** [review]

**Pull request merged**
